### PR TITLE
Add Bblfsh3 compatibility

### PR DESCRIPTION
**Pull request by Guillemdb**

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@Guillemdb Please rebase, I see merge commits

**Comment by Guillemdb** [comment]
@zurk Can you please confirm that the output of the failed tests is corrects and that the differences are due to the new bblfsh version? when you do, I will add the new output as the test values.

**Comment by Guillemdb** [comment]
I am afraid is not that simple, because setting the old version involves changing 20 files. The names are the same (changing string capitalization) but there are minor dfiferences. (the diferences are printed in the travis build, you can check them [here](https://travis-ci.com/src-d/ml-core/jobs/218222971#L765). 

I don't know how this is supposed to work in the first place, but to me the new outputs make sense.

File: sourced/ml/core/algorithms/uast_id_distance.py [hunk 1]
```diff
@@ -79,8 +81,9 @@ def _process_uast(self, uast: bblfsh.Node) -> Iterable:
             node, ancestors = stack.pop()
             yield from self._process_point(node, ancestors)
             ancestors = list(ancestors)
+
```
Comment (vmarkovtsev) [inline sourced/ml/core/algorithms/uast_id_distance.py]:
```suggestion
```

File: sourced/ml/core/algorithms/uast_id_distance.py [hunk 2]
```diff
@@ -105,18 +108,23 @@ class Uast2IdLineDistance(Uast2IdDistance):
     """
 
     def _process_uast(self, uast):
+
```
Comment (vmarkovtsev) [inline sourced/ml/core/algorithms/uast_id_distance.py]:
```suggestion
```

File: sourced/ml/core/algorithms/uast_id_distance.py [hunk 3]
```diff
@@ -105,18 +108,23 @@ class Uast2IdLineDistance(Uast2IdDistance):
...
         stack = [(uast, [0, 0])]
         while stack:
             node, last_position = stack.pop()
-            if node.start_position.line != 0:
+
```
Comment (vmarkovtsev) [inline sourced/ml/core/algorithms/uast_id_distance.py]:
```suggestion
```

File: sourced/ml/core/tests/test_bblfsh_utils.py [hunk 4]
```diff
@@ -76,9 +76,9 @@ def test_v240(self):
     def test_v250(self):
         self.assertTrue(self.__check_bblfsh_version_support("v2.5.0"), self.er_msg)
 
-    @classmethod
-    def tearDownClass(cls):
-        cls.docker_client.close()
+    # @classmethod
```
Comment (vmarkovtsev) [inline sourced/ml/core/tests/test_bblfsh_utils.py]: Why?

File: sourced/ml/core/utils/bblfsh.py [hunk 5]
```diff
@@ -1,19 +1,66 @@
 from distutils.version import StrictVersion
+from typing import List, Sequence
+
+from bblfsh.compat import CompatBblfshClient as BblfshClient
+from bblfsh.compat import CompatNodeIterator, Node
 
-from bblfsh.client import BblfshClient
 
 BBLFSH_VERSION_LOW = "2.2"
```
Comment (vmarkovtsev) [inline sourced/ml/core/utils/bblfsh.py]:
```suggestion
BBLFSH_VERSION_LOW = "3.0"
```

File: sourced/ml/core/utils/bblfsh.py [hunk 6]
```diff
@@ -1,19 +1,66 @@
...
+    :param host: bblfsh server host.
+    :param port: bblfsh server port.
+    :return: True if bblfsh version specified matches requirements.
     """
     # get version and remove leading 'v'
     version = StrictVersion(BblfshClient("%s:%s" % (host, port)).version().version.lstrip("v"))
     return StrictVersion(BBLFSH_VERSION_LOW) <= version < StrictVersion(BBLFSH_VERSION_HIGH)
+
+
+def iterate_children(nodes: Sequence[Node]) -> CompatNodeIterator:
```
Comment (vmarkovtsev) [inline sourced/ml/core/utils/bblfsh.py]: Let's have our `Node` class extension, with blackjack and whores.

The best would be to add these helper functions directly to the Node class in the Babelfish client. Can you please PR there.

File: sourced/ml/core/tests/test_bblfsh_utils.py (hunk 4 above)
Comment (Guillemdb) [reply sourced/ml/core/tests/test_bblfsh_utils.py]: I get this when I run it locally:
```
======================================================================
ERROR: tearDownClass (sourced.ml.core.tests.test_bblfsh_utils.BblfshUtilsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/guillem/srcd/ml-core/sourced/ml/core/tests/test_bblfsh_utils.py", line 81, in tearDownClass
    cls.docker_client.close()
  File "/usr/local/lib/python3.6/dist-packages/docker/client.py", line 196, in __getattr__
    raise AttributeError(' '.join(s))
AttributeError: 'DockerClient' object has no attribute 'close' In Docker SDK for Python 2.0, this method is now on the object APIClient. See the low-level API section of the documentation for more details.

```

**Comment by Guillemdb** [comment]
Update: I looked again at the test, and I was wrong. They do not make sense at all. I am not extracting the lines and the tokens correctly, but I am already working on it. Once I am sure I know how to replicate correctly the old API I will make the PR to bblfsh and an update on the community channel.

File: sourced/ml/core/utils/bblfsh.py (hunk 6 above)
Comment (Guillemdb) [reply sourced/ml/core/utils/bblfsh.py]: On halt till I'm sure I am replicating the old API correctly.

Comment (Guillemdb) [reply sourced/ml/core/utils/bblfsh.py]: @zurk All the breaking changes are handled using the functions in this file. This way you can quickly see in pycharm all the places where the new bblfsh version breaks.

File: sourced/ml/core/utils/bblfsh.py [hunk 7]
```diff
@@ -1,19 +1,66 @@
...
+    The node.children attribute is transformed to an iterator, and then filtered by a
+    CompatNodeIterator. The CompatNodeIterator transforms the dict nodes into Node instances.
+    """
+    def node_generator(node_seq):
+        for val in node_seq:
+            yield val
+    return CompatNodeIterator(node_generator(nodes))
+
+
+def get_node_tokens(node: Node) -> List[str]:
```
Comment (Guillemdb) [inline sourced/ml/core/utils/bblfsh.py]: This has a different signature than the original. I am not sure if `".".join(get_node_tokens(node))` would replicate the original one. (I don't know how it looked in the first place).

Unfortunatelly, the [`@token` field](https://docs.sourced.tech/babelfish/uast/semantic-uast#common-fields) describe here does not exists.

**Comment by zurk** [comment]
As was discussed I am taking over this PR to let @Guillemdb work on the more important stuff

**Comment by zurk** [comment]
Work continues in https://github.com/src-d/ml-core/pull/31

**Pull request closed without merging**